### ISSUE TITLE
Master records updated when related records have not changed

### ISSUE THREAD

**Issue by jondavis9898** [afawcett/declarative-lookup-rollup-summaries#226]
During change detection processing of trigger initiated rollups (result of an "update"), two situations are occurring that result in what is believed to be an unnecessary increase in governed limit operations (queryrows, dmlrecords).

1) When one (or more) records are detected as having a field that is involved in a rollup changed, all master record ids for all records in the trigger batch are evaluated and updated.

2) When the only change on a record is a re-parent, all master record ids for all rollups are evaluated and updated instead of just those master ids related to the re-parenting.

I'm not sure if this is intentional or not but it seems like the surface area should be minimized to only update master records that have a change associated to them.  As it stands, QueryRows and DMLRows Limits take quite a hit when there is only a single record change that effects just a single master record.  For example, if a group of 200 records were processed and if each record had a different "parent" and only one of the records had a change to a "rollup related field", all 200 master records would be updated instead of just the one that actually changed.

**Comment by afawcett** [comment]
Its a bit late a night now for me to be looking at this, will need to look again with less tired eyes

I was wondering though the change detection logic, is this now not shared between trigger code path and the new developer API code path?

**Comment by jondavis9898** [comment]
I saw that you merged in the new dev API ( PR #221) so yes, this problem is now shared between trigger code path and new developer API.  I hadn't finished the updates to the dev api tests yet (adding in the governor checks you requested) so hadn't planned on the dev api merge occurring yet.  My thinking was that we'd close the loop on this one first and then I'd merge that in to dev api change along with me updating the tests accounting for the fix from this Issue.

With the dev api merged in, I'll need to update PR #227 as well as add the governor checks on the dev api tests.

Short story is that because of this issue, my tests for dev api were encountering unexpected results so I didn't want to finish them without knowing what values to assert against (higher value if we don't address this issue, lower value if we do).

With the dev api merged in, I'll update PR #227 so you can evaluate after getting some rest :)

**Comment by afawcett** [comment]
Ok taken a look at this now... good spot!

Comparing with the earlier code....

			// Determine if a a field referenced on the lookup has changed and thus if the lookup itself needs recalculating
			Set<String> fieldsChanged = new Set<String>();  
			for(SObject childRecord : childRecords)
			{
				// Determine if any of the fields referenced on our selected rollups have changed on this record
				for(String fieldToSearch : fieldsToSearchForChanges)
				{
					SObject oldChildRecord = Trigger.oldMap.get(childRecord.Id);
					Object newValue = childRecord.get(fieldToSearch);
					Object oldValue = oldChildRecord.get(fieldToSearch);
					// Register this field as having changed?
					if(newValue != oldValue)
						fieldsChanged.add(fieldToSearch);
					// Add both old and new value to master record Id list for relationship fields to ensure old and new parent master records are updated (re-parenting)
					if(relationshipFields.contains(fieldToSearch))
					{
						if(newValue!=null)
							masterRecordIds.add((Id) newValue);
						if(oldValue!=null)
							masterRecordIds.add((Id) oldValue);		
					}
				}
			}

Would this not have fixed it?

			// Determine if a a field referenced on the lookup has changed and thus if the lookup itself needs recalculating
			Set<String> fieldsChanged = new Set<String>();  
			for(SObject childRecord : childRecords)
			{
				// Determine if any of the fields referenced on our selected rollups have changed on this record
				for(String fieldToSearch : fieldsToSearchForChanges)
				{
					SObject oldChildRecord = Trigger.oldMap.get(childRecord.Id);
					Object newValue = childRecord.get(fieldToSearch);
					Object oldValue = oldChildRecord.get(fieldToSearch);
					// Register this field as having changed?
					if(newValue != oldValue) 
					{
						fieldsChanged.add(fieldToSearch);
						// Add both old and new value to master record Id list for relationship fields to ensure old and new parent master records are updated (re-parenting)
						if(relationshipFields.contains(fieldToSearch))
						{
							if(newValue!=null)
								masterRecordIds.add((Id) newValue);
							if(oldValue!=null)
								masterRecordIds.add((Id) oldValue);		
						}
					}
				}
			}

**Comment by jondavis9898** [comment]
Unfortunately no.  When I first looked in to this, I thought the same thing (just move the relationshipfield check inside the change detection block).  However, this would end up checking relationship fields set for the field that is actually changing (fieldToSearch) and therefore we'd miss adding relationship ids in a number of cases.  For example, if "Amount__c" changed, "Amount__c" wouldn't be in relationship fields set so master id's wouldn't be added and therefore no records would end up being updated.  What needs to happen is to add master ids if any field changes but on a record-by-record basis, if that record hasn't changed, do not add master ids.

To do that, I separated the old/new checking since we need to iterate all non-relationship fields looking for any change before we can determine if we should add master ids.

1) Check for any non-relationship type field changes and if there are any, toggle helper variable to ensure add all master ids from all relationship fields to the master ids list.  This ensures that all master ids are eligible to be updated if a field in its respective rollup(s) changed.
2) Separately, check relationship fields for changes and if it has changed (or if any non-relationship field changed) add the old/new ids in to master ids.
